contracts/WakuRlnRegistry.sol

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import {WakuRln} from "./WakuRln.sol";
+import {IPoseidonHasher} from "rln-contract/PoseidonHasher.sol";
+import {Ownable} from "openzeppelin-contracts/contracts/access/Ownable.sol";
+import "forge-std/console.sol";
+
+error StorageAlreadyExists(address storageAddress);
+error NoStorageContractAvailable();
+error IncompatibleStorage();
+error IncompatibleStorageIndex();
+
+contract WakuRlnRegistry is Ownable {
+    uint16 public nextStorageIndex;
+    mapping(uint16 => address) public storages;
+
+    uint16 public usingStorageIndex = 0;
+
+    IPoseidonHasher public immutable poseidonHasher;
+
+    event NewStorageContract(uint16 index, address storageAddress);
+
+    constructor(address _poseidonHasher) Ownable() {
+        poseidonHasher = IPoseidonHasher(_poseidonHasher);
+    }
+
+    function _insertIntoStorageMap(address storageAddress) internal {
+        storages[nextStorageIndex] = storageAddress;
+        emit NewStorageContract(nextStorageIndex, storageAddress);
+        nextStorageIndex += 1;
+    }
+
+    function registerStorage(address storageAddress) external onlyOwner {
+        if (storages[nextStorageIndex] != address(0)) revert StorageAlreadyExists(storageAddress);
+        WakuRln wakuRln = WakuRln(storageAddress);
+        if (wakuRln.poseidonHasher() != poseidonHasher) revert IncompatibleStorage();
+        if (wakuRln.contractIndex() != nextStorageIndex) revert IncompatibleStorageIndex();
+        _insertIntoStorageMap(storageAddress);
+    }
+
+    function newStorage() external onlyOwner {
+        WakuRln newStorageContract = new WakuRln(address(poseidonHasher), nextStorageIndex);
+        _insertIntoStorageMap(address(newStorageContract));
+    }
+
+    function register(uint256[] calldata commitments) external payable {
+        if (usingStorageIndex >= nextStorageIndex) revert NoStorageContractAvailable();
+
+        // iteratively check if the storage contract is full, and increment the usingStorageIndex if it is
+        while (true) {
+            try WakuRln(storages[usingStorageIndex]).register(commitments) {
+                break;
+            } catch (bytes memory err) {
+                if (keccak256(err) != keccak256(abi.encodeWithSignature("FullTree()"))) {
+                    assembly {
+                        revert(add(32, err), mload(err))
+                    }
+                }
```

Comment:
Correct me if I'm wrong, but based on `WakuRln.register()` as of this commit, it looks like this will never revert with `error FullTree()` because it doesn't seem to inherit from `RlnBase`s `register()`:

https://github.com/waku-org/waku-rln-contract/blob/ace937c3adfe2617dfa75e907325055594e307b5/contracts/WakuRln.sol#L19-L26

Am I missing something here? This looks like this could be running out of gas otherwise...

Reply:
Geez, i completely missed those error cases - will fix, thanks!

Reply:
Addressed in [ae01148](https://github.com/waku-org/waku-rln-contract/pull/3/commits/ae01148ffc151ec4d1e0047b84fdc5bb61aa41eb)